bdk_chain/src/keychain/keychain_txout_index.rs

```diff
@@ -156,11 +158,40 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
             .map(|((_, derivation_index), spk)| (*derivation_index, spk))
     }
 
-    /// Get the derivation index after the current one
-    pub fn next_derivation_index(&self, keychain: &K) -> u32 {
-        self.derivation_index(keychain)
-            .map(|index| index + 1)
-            .unwrap_or(0)
+    /// Get the derivation index after the current one, unless the descriptor has no wildcards.
+    ///
+    /// Return `true` if the `next_index` needs to be derived.
+    /// Return `false` if the `next_index` is already derived and cannot be increased further.
+    /// for example, in cases of:
+    ///  - non derivable descriptor that already has one `script_pubkey` stored in the tracker.
+    ///  - derivable descriptor when `script_pubkey` index hits numeric limit.
+    ///
+    /// Note that, using `scripts_pubkey`s from such cases can lead to address reuse.
+    pub fn next_derivation_index(&self, keychain: &K) -> (u32, bool) {
+        // we can only get the next index if wildcard exists
+        let has_wildcard = self
+            .keychains
+            .get(keychain)
+            .expect("Invalid keychain: keychain doesn't exist in tracker")
+            .has_wildcard();
+
+        match self.derivation_index(keychain) {
+            // for all descriptors, if there is no index, next_index is always 0
+            None => (0, true),
+            // for derivable descriptors, we get the next index (without overflow)
+            // for un-derivable descriptors we always get next_index = 0
+            Some(index) => {
+                if !has_wildcard {
+                    (0, false)
+                } else {
+                    if let Some(new_index) = index.checked_add(1) {
```

Comment:
Do we also need to check bounds at `2^31`? It might be more convenient to create a constant.

Reply:
Done in d488f4eb250de16074ef8c0b44acb1a64a28ef06